keps/sig-network/0015-20180614-SCTP-support.md

```diff
@@ -0,0 +1,291 @@
+```
+apiVersion: networking.k8s.io/v1
+kind: NetworkPolicy
+metadata:
+  name: myservice-network-policy
+  namespace: myapp
+spec:
+  podSelector:
+    matchLabels:
+      role: myservice
+  policyTypes:
+  - Ingress
+  - Egress
+  ingress:
+  - from:
+    - ipBlock:
+        cidr: 172.17.0.0/16
+        except:
+        - 172.17.1.0/24
+    - namespaceSelector:
+        matchLabels:
+          project: myproject
+    - podSelector:
+        matchLabels:
+          role: myclient
+    ports:
+    - protocol: SCTP
+      port: 7777
+```
+#### User space SCTP stack
+As a user of Kubernetes I want to deploy and run my applications that use a user space SCTP stack. 
+
+### Implementation Details/Notes/Constraints [optional]
+
+<!---
+What are the caveats to the implementation?
+What are some important details that didn't come across above.
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they releate.
+--->
+#### SCTP in Services
+The Kubernetes API modification for Services is obvious.
+The selected port shall be reserved on the node, just like for TCP and UDP now. Unfortunately, golang does not have native SCTP support in the "net" package, so in order to access the kernel's SCTP API we have to introduce a new 3rd party package as a new vendor package.
+For Services with type=LoadBalancer we have to check how the cloud provider implementations handle new protocols, and we have to make sure that if SCTP is not supported then the request for a new load balancer, firewall rule, etc. with protocol=SCTP is rejected gracefully.
+
+#### SCTP in NetworkPolicy
+The Kubernetes API modification for the NetworkPolicy is obvious.
+In order to utilize the new protocol value the network controller must support it. 
+
+#### Interworking with applications that use a user space SCTP stack
+A userspace SCTP stack implementation cannot work together with the SCTP kernel module (lksctp) on the same node. That is, the loading of the SCTP kernel module must be avoided on nodes where such applications that use userspace SCTP stack are planned to be run. The problem comes with the introduction of the SCTP protocol option for Services with Virtual IP: once such a service is created the relevant iptables/ipvs management logic kicks-in on every node, and as a consequence it loads the SCTP kernel module. There are some ideas how to solve this interworking problem:
+
+1. "-p sctp" is not used in the iptables rules, the processing of requests to the Virtual IP is executed purely based on the destination IP address. In case of ipvs the protocol is a mandatory parameter, so ipvs with SCTP rules cannot be used on the node where userspace SCTP applications should run.
+2. Fall back to the user space proxy on those specific nodes. The user space proxy shall also use a user space SCTP stack, of course. Also the iptables rules that direct the client traffic to the userspace proxy must be created without the "-p sctp" option.
```

Comment:
"of course"? Ah... is it not possible to implement a generic SCTP proxy via the sockets API?

If the userspace proxy needs to use the userspace SCTP stack then that means the "tainting" runs in both directions: if you want to run the userspace proxy on a node for whatever reason, then you can't run containers that expect to use kernel SCTP on that node.

Reply:
And of course if you don't use kernel SCTP support then that kind of kills my earlier claim that NetworkPolicy would be easy to implement because the kernel does all the hard stuff. Pluging with iptables-, OVS-, or eBPF-based NetworkPolicy presumably would not be able to do SCTP at all on nodes without kernel SCTP.

Reply:
@thockin Indeed, a user space proxy does not sound good, but on a node where a userspace SCTP app runs we cannot use the SCTP kernel module anyway. So, if we want to have a proxy on those nodes, then we have to implement the SCTP stack for that proxy in the userspace.
@danwinship Exactly, you see it right. On a node where a userspace SCTP application runs we cannot run such applications that would require the SCTP kernel module. And it is true on the other way around, too.

Reply:
Thinking about this part again: it will not work. In order to route a request to e.g. a userspace proxy the port on which the userspace proxy listens must be configured in iptables/ipvs. And it means that the protocol must be configured, that loads the kernel module.
Currently I think, there is no nice way to solve the interworking of userspace SCTP stacks and kernel module based services in the same k8s cluster.